server/src/core/api-key/api-key.resolver.ts

```diff
@@ -0,0 +1,95 @@
+import { Args, Mutation, Query, Resolver } from '@nestjs/graphql';
+import { NotFoundException, UseGuards } from '@nestjs/common';
+
+import { accessibleBy } from '@casl/prisma';
+import { v4 } from 'uuid';
+
+import { AbilityGuard } from 'src/guards/ability.guard';
+import { AuthWorkspace } from 'src/decorators/auth-workspace.decorator';
+import { Workspace } from 'src/core/@generated/workspace/workspace.model';
+import { CreateOneApiKeyArgs } from 'src/core/@generated/api-key/create-one-api-key.args';
+import { ApiKey } from 'src/core/@generated/api-key/api-key.model';
+import { FindManyApiKeyArgs } from 'src/core/@generated/api-key/find-many-api-key.args';
+import { DeleteOneApiKeyArgs } from 'src/core/@generated/api-key/delete-one-api-key.args';
+import { CheckAbilities } from 'src/decorators/check-abilities.decorator';
+import {
+  CreateApiKeyAbilityHandler,
+  DeleteApiKeyAbilityHandler,
+  ReadApiKeyAbilityHandler,
+} from 'src/ability/handlers/api-key.ability-handler';
+import { JwtAuthGuard } from 'src/guards/jwt.auth.guard';
+import { UserAbility } from 'src/decorators/user-ability.decorator';
+import { AppAbility } from 'src/ability/ability.factory';
+import { TokenService } from 'src/core/auth/services/token.service';
+
+import { ApiKeyService } from './api-key.service';
+
+@UseGuards(JwtAuthGuard)
+@Resolver(() => ApiKey)
+export class ApiKeyResolver {
+  constructor(
+    private readonly apiKeyService: ApiKeyService,
+    private readonly tokenService: TokenService,
+  ) {}
+
+  @Mutation(() => ApiKey)
+  @UseGuards(AbilityGuard)
+  @CheckAbilities(CreateApiKeyAbilityHandler)
+  async createOneApiKey(
+    @Args() args: CreateOneApiKeyArgs,
+    @AuthWorkspace() { id: workspaceId }: Workspace,
+  ): Promise<Partial<ApiKey>> {
+    const apiKeyId = v4();
+    const customApiKey = await this.tokenService.generateApiKeyToken(
+      workspaceId,
+      apiKeyId,
+      args.data.expiresAt,
+    );
+    await this.apiKeyService.create({
+      data: {
+        id: apiKeyId,
+        key: customApiKey.token,
```

Comment:
@martmull You shouldn't store the token inside the database, but should do the same as for refresh-token.
I mean by that, creating the model inside the database, and adding the id of the model inside the token payload.

```typescript
  async generateRefreshToken(userId: string): Promise<AuthToken> {
    const secret = this.environmentService.getRefreshTokenSecret();
    const expiresIn = this.environmentService.getRefreshTokenExpiresIn();
    assert(expiresIn, '', InternalServerErrorException);
    const expiresAt = addMilliseconds(new Date().getTime(), ms(expiresIn));

    const refreshTokenPayload = {
      userId,
      expiresAt,
    };
    const jwtPayload = {
      sub: userId,
    };

    const refreshToken = await this.prismaService.client.refreshToken.create({
      data: refreshTokenPayload,
    });

    return {
      token: this.jwtService.sign(jwtPayload, {
        secret,
        expiresIn,
        // Jwtid will be used to link RefreshToken entity to this token
        jwtid: refreshToken.id,
      }),
      expiresAt,
    };
  }
```

Reply:
I was ok with that, but when we discussed with @charlesBochet on Monday and it has been decided that the token should be retrievable by the user. So we need to store it then

Reply:
seen with Charles, lets remove tokens from db